<a href="https://colab.research.google.com/github/paweenatoei/KKU_data_mining/blob/master/Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Project Data Mining
#เรื่อง Association
#หาความสัมพันธ์ของคะแนนภาพยนต์กับนักแสดง ว่ามีความสัมพันธ์ต่อกันหรือไม่ นักแสดงจะส่งผลต่อคะแนนของภาพยนต์หรือไม่

###วัตถุประสงค์
> เพื่อดูว่านักแสดงมีผลต่อคะแนนของภาพยนต์ไหม


> เพื่อเป็นประโยชน์ในการผลิตภาพยนต์

###แหล่งที่มาของข้อมูล
เว็บ Kaggle
> https://www.kaggle.com/tmdb/tmdb-movie-metadata?select=tmdb_5000_movies.csv&fbclid=IwAR2Jb24ba0XhnV3q_XHFQQx7_LNr14k0BKGEu6BqBVkirjYD3jGMK8ySf1U






##การนำเข้าข้อมูล

In [ ]:
import pandas as pd
import os
#เป็นการนำเข้าตัวแปรเพื่อใช้ในการรันข้อมูล

In [ ]:
from google.colab import drive
drive.mount('/content/drive')
#เป็นการนำเข้าไฟล์ข้อมูลจากไดร์ฟมายัง colab

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
path = '/content/drive/My Drive/TB_pro'
#path คือ การกำหนดตัวแปรให้เป็นที่อยู่ของไฟล์

In [ ]:
tmdb_5000_movies = os.path.join(path,'tmdb_5000_movies.csv')
#สร้างที่อยู่ของข้อมูลชุดที่ 1 โดยใช้ os.path.join

In [ ]:
tmdb_5000_movies_df = pd.read_csv(tmdb_5000_movies)
tmdb_5000_movies_df
#แสดงไฟล์ข้อมูลชุดที่ 1

budget  ... vote_count
0     237000000  ...      11800
1     300000000  ...       4500
2     245000000  ...       4466
3     250000000  ...       9106
4     260000000  ...       2124
...         ...  ...        ...
4798     220000  ...        238
4799       9000  ...          5
4800          0  ...          6
4801          0  ...          7
4802          0  ...         16

[4803 rows x 20 columns]

In [ ]:
tmdb_5000_movies_df.isnull().any()
#เป็นการตรวจสอบว่าของมูลชุดที่ 1 ในแต่ละคอลัมมีข้อมูลที่ขาดหายไหม
#ถ้า false คือ ไม่มีข้อมูลที่ขาดหาย
#true คือ มีข้อมูลที่ขาดหาย

budget                  False
genres                  False
homepage                 True
movie_id                False
keywords                False
original_language       False
original_title          False
overview                 True
popularity              False
production_companies    False
production_countries    False
release_date             True
revenue                 False
runtime                  True
spoken_languages        False
status                  False
tagline                  True
title                   False
vote_average            False
vote_count              False
dtype: bool

In [ ]:
tmdb_5000_credits = os.path.join(path,'tmdb_5000_credits.csv')
#สร้างที่อยู่ของข้อมูลชุดที่ 2 โดยใช้ os.path.join

In [ ]:
tmdb_5000_credits_df = pd.read_csv(tmdb_5000_credits)
tmdb_5000_credits_df
#แสดงไฟล์ข้อมูลชุดที่ 2

movie_id  ...                                               crew
0        19995  ...  [{"credit_id": "52fe48009251416c750aca23", "de...
1          285  ...  [{"credit_id": "52fe4232c3a36847f800b579", "de...
2       206647  ...  [{"credit_id": "54805967c3a36829b5002c41", "de...
3        49026  ...  [{"credit_id": "52fe4781c3a36847f81398c3", "de...
4        49529  ...  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...
...        ...  ...                                                ...
4798      9367  ...  [{"credit_id": "52fe44eec3a36847f80b280b", "de...
4799     72766  ...  [{"credit_id": "52fe487dc3a368484e0fb013", "de...
4800    231617  ...  [{"credit_id": "52fe4df3c3a36847f8275ecf", "de...
4801    126186  ...  [{"credit_id": "52fe4ad9c3a368484e16a36b", "de...
4802     25975  ...  [{"credit_id": "58ce021b9251415a390165d9", "de...

[4803 rows x 4 columns]

###เนื่องจากจะใช้คอลัมน์ cast หาความสัมพันธ์ ซึ่งคอลัมน์ cast มีเนื้อหาที่เป็นขยะ เช่น รหัสนักแสดง เพศ แสดงเป็นใคร เป็นต้น ซึ่งเราต้องการแค่ชื่อนักแสดง จึงทำการแยกเนื้อหาในคอลัมน์ออกใน excel โดยใช้คำสั่ง Text to Columns และใช้ : เป็นตัวแยก แล้วทำการลบเนื้อหาที่เป็นขยะและเหลือไว้แค่ชื่อนักแสดงด้วยมือ (ซึ่งนำรายชื่อนักแสดงมา 5 คนที่เป็นนักแสดงนำ เพื่อนำมาเป็นตัวอย่างในการหาความสัมพันธ์ เนื่องจากภาพยนต์แต่ละเรื่องมีนักแสดงเป็นจำนวนมาก)

In [ ]:
tmdb_5000_credits_11 = os.path.join(path,'tmdb_5000_credits_11.csv')
#สร้างที่อยู่ของข้อมูลชุดที่ 2(ใหม่) โดยใช้ os.path.join

In [ ]:
tmdb_5000_credits_11_df = pd.read_csv(tmdb_5000_credits_11)
tmdb_5000_credits_11_df
#แสดงไฟล์ข้อมูลชุดที่ 2(ใหม่)
#โดยจะใช้ข้อมูลชุดที่ 2 ชุดนี้เพื่อนำไปหาผลรัน

movie_id  ...                                               crew
0        19995  ...  [{"credit_id": "52fe48009251416c750aca23", "de...
1          285  ...  [{"credit_id": "52fe4232c3a36847f800b579", "de...
2       206647  ...  [{"credit_id": "54805967c3a36829b5002c41", "de...
3        49026  ...  [{"credit_id": "52fe4781c3a36847f81398c3", "de...
4        49529  ...  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...
...        ...  ...                                                ...
4798      9367  ...  [{"credit_id": "52fe44eec3a36847f80b280b", "de...
4799     72766  ...  [{"credit_id": "52fe487dc3a368484e0fb013", "de...
4800    231617  ...  [{"credit_id": "52fe4df3c3a36847f8275ecf", "de...
4801    126186  ...  [{"credit_id": "52fe4ad9c3a368484e16a36b", "de...
4802     25975  ...  [{"credit_id": "58ce021b9251415a390165d9", "de...

[4803 rows x 9 columns]

In [ ]:
tmdb_5000_credits_11_df.isnull().any()
#เป็นการตรวจสอบว่าของมูลชุดที่ 2 ในแต่ละคอลัมมีข้อมูลที่ขาดหายไหม
#ถ้า false คือ ไม่มีข้อมูลที่ขาดหาย
#true คือ มีข้อมูลที่ขาดหาย

movie_id    False
title       False
cast1        True
cast2        True
cast3        True
cast4        True
cast5        True
cast_new     True
crew         True
dtype: bool

In [ ]:
merge_tb = tmdb_5000_movies_df.merge(tmdb_5000_credits_11_df,how='left', left_on='movie_id',right_on='movie_id')
merge_tb
#เป็นการเชื่อมตาราง โดยใช้คำสั่ง merge และให้ movie_id เป็นตัวเชื่อมระหว่างข้อมูล 2 ชุด

budget  ...                                               crew
0     237000000  ...  [{"credit_id": "52fe48009251416c750aca23", "de...
1     300000000  ...  [{"credit_id": "52fe4232c3a36847f800b579", "de...
2     245000000  ...  [{"credit_id": "54805967c3a36829b5002c41", "de...
3     250000000  ...  [{"credit_id": "52fe4781c3a36847f81398c3", "de...
4     260000000  ...  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...
...         ...  ...                                                ...
4798     220000  ...  [{"credit_id": "52fe44eec3a36847f80b280b", "de...
4799       9000  ...  [{"credit_id": "52fe487dc3a368484e0fb013", "de...
4800          0  ...  [{"credit_id": "52fe4df3c3a36847f8275ecf", "de...
4801          0  ...  [{"credit_id": "52fe4ad9c3a368484e16a36b", "de...
4802          0  ...  [{"credit_id": "58ce021b9251415a390165d9", "de...

[4803 rows x 28 columns]

###เป็นการสร้างตารางใหม่ตามที่จะใช้ในการหาความสัมพันธ์

In [ ]:
merge_tb[['vote_average','original_title','cast_new']]
#สร้างตารางใหม่เพื่อให้เหลือเฉพาะคอลัมน์ที่สนใจ

vote_average  ...                                           cast_new
0              7.2  ...   "Sam Worthington", "Zoe Saldana", "Stephen La...
1              6.9  ...   "Johnny Depp", "Orlando Bloom", "Stellan Skar...
2              6.3  ...   "Daniel Craig", "Christoph Waltz", "Ralph Fie...
3              7.6  ...   "Christian Bale", "Michael Caine", "Anne Hath...
4              6.1  ...   "Taylor Kitsch", "Lynn Collins", "Willem Dafo...
...            ...  ...                                                ...
4798           6.6  ...   "Carlos Gallardo", "Jaime de Hoyos", "Peter M...
4799           5.9  ...   "Edward Burns", "Kerry Bish\u00e9", "Marsha D...
4800           7.0  ...   "Eric Mabius", "Kristin Booth", "Crystal Lowe...
4801           5.7  ...   "Daniel Henney", "Eliza Coupe", "Bill Paxton"...
4802           6.3  ...   "Drew Barrymore", "Brian Herzlinger", "Corey ...

[4803 rows x 3 columns]

#Drop missing value

เป็นการดรอป missing

In [ ]:
!pip install dp

In [ ]:
merge_tb_dp = merge_tb.dropna()
merge_tb_dp
#เป็นการลบข้อมูลที่ missing ออก

budget  ...                                               crew
0     237000000  ...  [{"credit_id": "52fe48009251416c750aca23", "de...
1     300000000  ...  [{"credit_id": "52fe4232c3a36847f800b579", "de...
2     245000000  ...  [{"credit_id": "54805967c3a36829b5002c41", "de...
3     250000000  ...  [{"credit_id": "52fe4781c3a36847f81398c3", "de...
4     260000000  ...  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...
...         ...  ...                                                ...
4772      31192  ...  [{"credit_id": "52fe45f0c3a36847f80e3327", "de...
4773      27000  ...  [{"credit_id": "52fe434ac3a36847f8049253", "de...
4781      22000  ...  [{"credit_id": "52fe4ee39251416c911412e7", "de...
4796       7000  ...  [{"credit_id": "52fe45e79251416c75066791", "de...
4801          0  ...  [{"credit_id": "52fe4ad9c3a368484e16a36b", "de...

[1461 rows x 28 columns]

In [ ]:
merge_tb_dp.isnull().any()
#หลังจากที่ทำการลบ missing ออกแล้ว จะทำการตรวจสอบว่าจากที่ทำการลบไปยังมีข้อมูลที่ missing อยู่ไหม
#false คือ ไม่มีข้อมูล missing
#true คือ มีข้อมูลที่ missing
#จากการตรวจสอบพบว่าไม่มีข้อมูลที่ missing แล้ว

budget                  False
genres                  False
homepage                False
movie_id                False
keywords                False
original_language       False
original_title          False
overview                False
popularity              False
production_companies    False
production_countries    False
release_date            False
revenue                 False
runtime                 False
spoken_languages        False
status                  False
tagline                 False
title_x                 False
vote_average            False
vote_count              False
title_y                 False
cast1                   False
cast2                   False
cast3                   False
cast4                   False
cast5                   False
cast_new                False
crew                    False
dtype: bool

#Percent of missing data from dropna

##เป็นการตรวจสอบว่ามีข้อมูล missing กี่ % ของข้อมูลทั้งหมด

In [ ]:
size_before_drop = 4803
size_after_drop = 1493
print(f'size before drop = {size_before_drop} \nsize after drop = {size_after_drop}\nwhich is {100*((size_before_drop-size_after_drop)/size_before_drop)}')
#จากการตรวจสอบพบว่ามีข้อมูล missing อยู่ 68.92%

size before drop = 4803 
size after drop = 1493
which is 68.91526129502394


#Data Mining


##จากนั้นจะแบ่งกลุ่มภาพยนต์ออกเป็น 2 กลุ่ม
###กลุ่มที่ 1 คือกลุ่มภาพยนต์ที่มีคะแนนน้อยกว่าหรือเท่ากับ 5
###กลุ่มที่ 2 คือกลุ่มภาพยนต์ที่มีคะแนนมากว่า 5

In [ ]:
table2 = merge_tb_dp[merge_tb_dp['vote_average']<= 5]
table2
#เป็นการแสดงภาพยนค์ที่มีคะแนนโหวตน้อยว่าหรือเท่ากับ 5

budget  ...                                               crew
91    165000000  ...  [{"credit_id": "5763d39cc3a36860a5000168", "de...
138   150000000  ...  [{"credit_id": "52fe433f9251416c750092c9", "de...
226   120000000  ...  [{"credit_id": "52fe4620c3a36847f80ee48d", "de...
242   120000000  ...  [{"credit_id": "566e93a9925141664f00eb4c", "de...
292   100000000  ...  [{"credit_id": "52fe435ac3a36847f804de45", "de...
...         ...  ...                                                ...
4569          0  ...  [{"credit_id": "52fe4d7b9251416c9111797f", "de...
4647     250000  ...  [{"credit_id": "586ace83c3a3686e2a0034fa", "de...
4701     150000  ...  [{"credit_id": "52fe4424c3a36847f8084a63", "de...
4723     100000  ...  [{"credit_id": "54951e769251413af7001c60", "de...
4765          0  ...  [{"credit_id": "52fe4ecf9251416c75163381", "de...

[111 rows x 28 columns]

In [ ]:
table3 = merge_tb_dp[merge_tb_dp['vote_average']> 5]
table3
#เป็นการแสดงภาพยนค์ที่มีคะแนนโหวตมากกว่า 5

budget  ...                                               crew
0     237000000  ...  [{"credit_id": "52fe48009251416c750aca23", "de...
1     300000000  ...  [{"credit_id": "52fe4232c3a36847f800b579", "de...
2     245000000  ...  [{"credit_id": "54805967c3a36829b5002c41", "de...
3     250000000  ...  [{"credit_id": "52fe4781c3a36847f81398c3", "de...
4     260000000  ...  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...
...         ...  ...                                                ...
4772      31192  ...  [{"credit_id": "52fe45f0c3a36847f80e3327", "de...
4773      27000  ...  [{"credit_id": "52fe434ac3a36847f8049253", "de...
4781      22000  ...  [{"credit_id": "52fe4ee39251416c911412e7", "de...
4796       7000  ...  [{"credit_id": "52fe45e79251416c75066791", "de...
4801          0  ...  [{"credit_id": "52fe4ad9c3a368484e16a36b", "de...

[1350 rows x 28 columns]

###จากนั้นจะเป็นการแสดงเฉพาะข้อมูลที่เราต้องการนำไปหาความสัมพันธ์

In [ ]:
table7 = table2.groupby('vote_average')['cast_new'].apply(list)
table7
#เป็นการแสดงข้อมูลของภาพยนต์ที่มีคะแนนน้อยกว่าหรือเท่ากับ 5 ว่ามีนักแสดงชื่ออะไรบ้าง

vote_average
0.0    [ "Paul Ben-Victor", "Ed Asner", "Jordenn Thom...
1.9    [ "Jon Voight", "Scott Baio", "Vanessa Angel",...
2.3    [ "Charlie Sheen", "Wayne Brady", "Hilary Duff...
3.0    [ "Matt Lanter", "Vanessa Lachey", "Nicole Ari...
3.2    [ "Daniel Tay", "Jimmy Fallon", "Jon Stewart",...
3.3    [ "Dorothy Stratten", "Stephen Macht", "Avery ...
3.5    [ "Ben Affleck", "Jennifer Lopez", "Justin Bar...
3.7    [ "Sean Maguire", "Carmen Electra", "Ken Davit...
3.8    [ "Ian Ziering", "Tara Reid", "Cassandra Scerb...
3.9    [ "Erin O'Brien", "Miesha Tate", "Holly Holm",...
4.1    [ "Adam Sandler", "Katie Holmes", "Al Pacino",...
4.2    [ "Chris Riggi", "Matt Lanter", "Jenn Proske",...
4.3    [ "Mike Myers", "Jessica Alba", "Justin Timber...
4.4    [ "Miles Teller", "Kate Mara", "Michael B. Jor...
4.5    [ "Tim Allen", "Kristin Davis", "Danny Glover"...
4.6    [ "George Carlin", "Andy Dick", "Sarah Michell...
4.7    [ "Noah Ringer", "Nicola Peltz", "Jackson Rath...
4.8    [ "Heike Ma

In [ ]:
table6 = table3.groupby('vote_average')['cast_new'].apply(list)
table6
#เป็นการแสดงข้อมูลของภาพยนต์ที่มีคะแนนมากกว่า 5 ว่ามีนักแสดงชื่ออะไรบ้าง

vote_average
5.1    [ "Ryan Reynolds", "Blake Lively", "Peter Sars...
5.2    [ "Mila Kunis", "Channing Tatum", "Sean Bean",...
5.3    [ "Steve Carell", "Lauren Graham", "John Goodm...
5.4    [ "Jeff Bridges", "Ryan Reynolds", "Kevin Baco...
5.5    [ "Nicholas Hoult", "Eleanor Tomlinson", "Ewan...
5.6    [ "John Cusack", "Amanda Peet", "Chiwetel Ejio...
5.7    [ "Ben Affleck", "Henry Cavill", "Amy Adams", ...
5.8    [ "Dakota Blue Richards", "Nicole Kidman", "Sa...
5.9    [ "Tobey Maguire", "Kirsten Dunst", "Thomas Ha...
6.0    [ "Shia LaBeouf", "Megan Fox", "Josh Duhamel",...
6.1    [ "Taylor Kitsch", "Lynn Collins", "Willem Daf...
6.2    [ "Will Smith", "Tommy Lee Jones", "Michael St...
6.3    [ "Daniel Craig", "Christoph Waltz", "Ralph Fi...
6.4    [ "Johnny Depp", "Pen\u00e9lope Cruz", "Kevin ...
6.5    [ "Henry Cavill", "Amy Adams", "Kevin Costner"...
6.6    [ "Raymond Ochoa", "Jack Bright", "Jeffrey Wri...
6.7    [ "Tobey Maguire", "Kirsten Dunst", "James Fra...
6.8    [ "Robert D

In [ ]:
!pip install apyori
#เป็นการติดตั้ง apyori เพื่อใช้ในการหาความสัมพันธ์

  Created wheel for apyori: filename=apyori-1.1.2-cp36-none-any.whl size=5975 sha256=1ea857f7d462c884b4f429320808bb54e4f929c38dfc7538b362ad22ac3bf305
  Stored in directory: /root/.cache/pip/wheels/5d/92/bb/474bbadbc8c0062b9eb168f69982a0443263f8ab1711a8cad0
Successfully built apyori


In [ ]:
from apyori import apriori

In [ ]:
watch = list(apriori(table6, min_support = 0.042 , min_confidence = 0.05))
watch 
#เป็นหาความสัมพันธ์ของนักแสดงกับกลุ่มภาพยนต์ที่คะแนนมากกว่า 5

[RelationRecord(items=frozenset({' "Elijah Wood", "Ian McKellen", "Viggo Mortensen", "Liv Tyler", "Orlando Bloom"'}), support=0.058823529411764705, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({' "Elijah Wood", "Ian McKellen", "Viggo Mortensen", "Liv Tyler", "Orlando Bloom"'}), confidence=0.058823529411764705, lift=1.0)]),
 RelationRecord(items=frozenset({' "Ewan McGregor", "Natalie Portman", "Hayden Christensen", "Ian McDiarmid", "Samuel L. Jackson"'}), support=0.058823529411764705, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({' "Ewan McGregor", "Natalie Portman", "Hayden Christensen", "Ian McDiarmid", "Samuel L. Jackson"'}), confidence=0.058823529411764705, lift=1.0)]),
 RelationRecord(items=frozenset({' "Jennifer Lawrence", "Josh Hutcherson", "Liam Hemsworth", "Woody Harrelson", "Elizabeth Banks"'}), support=0.08823529411764706, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'

#Summary
###จากการทำ Association Rule จะเห็นได้ว่า
ภาพยนตร์ที่มีคะแนนมากกว่า 5 จากคะแนนเต็ม 10

ที่มีรายชื่อนักแสดงดังนี้

> -"Elijah Wood", "Ian McKellen", "Viggo Mortensen", "Liv Tyler", "Orlando Bloom" มีผลต่อคะแนนของภาพยนต์ 5.88%

> -"Jennifer Lawrence", "Josh Hutcherson", "Liam Hemsworth", "Woody Harrelson", "Elizabeth Banks" มีผลต่อคะแนนของภาพยนต์ 8.82%

> สรุปได้ว่า นักแสดงไม่มีความสัมพันธ์กับคะแนนของภาพยนต์ หรือไม่มีผลต่อคะแนนของภาพยนต์ โดยนักแสดงไม่ได้ทำให้คะแนนของภาพยนต์มากขึ้น หรือน้อยลง แต่คะแนนของภาพยนต์ที่มากหรือน้อยลง อาจขึ้นอยู่กับคุณภาพ และเนื้อเรื่องของภาพยนต์ว่ามีคุณภาพที่ดี หรือเนื้อเรื่องที่ดี และน่าสนใจหรือไม่  










  

###สมาชิกในกลุ่ม
1. นางสาวจิรัชญา ฟองฟูม 613020038-0
2. นางสาววัชราภรณ์ บุญจวง 613020615-8
3. นางสาวปวีณา ตรีทศ 613021095-3